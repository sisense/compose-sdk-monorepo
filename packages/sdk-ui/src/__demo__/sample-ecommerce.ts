import { measureFactory } from '@sisense/sdk-data';

import { Commerce, Brand, Country, Category, DataSource } from './sample-ecommerce-autogenerated';

export { Commerce, Brand, Category, Country, DataSource };

/** adding predefined set of measures */
export const Measures = {
  SumCost: measureFactory.sum(Commerce.Cost).format('0,0$'),
  AvgCost: measureFactory.average(Commerce.Cost).format('0,0$'),
  SumRevenue: measureFactory.sum(Commerce.Revenue, 'Total Revenue').format('0,0$'),
  Quantity: measureFactory.sum(Commerce.Quantity, 'Total Quantity').format('0,0'),
  CountBrands: measureFactory.countDistinct(Commerce.BrandID).format('0,0'),
};
