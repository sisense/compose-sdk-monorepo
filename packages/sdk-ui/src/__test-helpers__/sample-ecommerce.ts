import {
  Dimension,
  DateDimension,
  Attribute,
  createAttribute,
  createDateDimension,
  createDimension,
} from '@sisense/sdk-data';

export const DataSource = 'Sample ECommerce';

interface BrandDimension extends Dimension {
  Brand: Attribute;
  BrandID: Attribute;
}
export const Brand = createDimension({
  name: 'Brand',
  Brand: createAttribute({
    name: 'Brand',
    type: 'text-attribute',
    expression: '[Brand.Brand]',
  }),
  BrandID: createAttribute({
    name: 'BrandID',
    type: 'numeric-attribute',
    expression: '[Brand.Brand ID]',
  }),
}) as BrandDimension;

interface CategoryDimension extends Dimension {
  Category: Attribute;
  CategoryID: Attribute;
}
export const Category = createDimension({
  name: 'Category',
  Category: createAttribute({
    name: 'Category',
    type: 'text-attribute',
    expression: '[Category.Category]',
  }),
  CategoryID: createAttribute({
    name: 'CategoryID',
    type: 'numeric-attribute',
    expression: '[Category.Category ID]',
  }),
}) as CategoryDimension;

interface CommerceDimension extends Dimension {
  AgeRange: Attribute;
  BrandID: Attribute;
  CategoryID: Attribute;
  Condition: Attribute;
  Cost: Attribute;
  CountryID: Attribute;
  DateMonth: Attribute;
  Gender: Attribute;
  Quantity: Attribute;
  Revenue: Attribute;
  VisitID: Attribute;
  Date: DateDimension;
}
export const Commerce = createDimension({
  name: 'Commerce',
  AgeRange: createAttribute({
    name: 'AgeRange',
    type: 'text-attribute',
    expression: '[Commerce.Age Range]',
  }),
  BrandID: createAttribute({
    name: 'BrandID',
    type: 'numeric-attribute',
    expression: '[Commerce.Brand ID]',
  }),
  CategoryID: createAttribute({
    name: 'CategoryID',
    type: 'numeric-attribute',
    expression: '[Commerce.Category ID]',
  }),
  Condition: createAttribute({
    name: 'Condition',
    type: 'text-attribute',
    expression: '[Commerce.Condition]',
  }),
  Cost: createAttribute({
    name: 'Cost',
    type: 'numeric-attribute',
    expression: '[Commerce.Cost]',
  }),
  CountryID: createAttribute({
    name: 'CountryID',
    type: 'numeric-attribute',
    expression: '[Commerce.Country ID]',
  }),
  DateMonth: createAttribute({
    name: 'DateMonth',
    type: 'numeric-attribute',
    expression: '[Commerce.Date (Month)]',
  }),
  Gender: createAttribute({
    name: 'Gender',
    type: 'text-attribute',
    expression: '[Commerce.Gender]',
  }),
  Quantity: createAttribute({
    name: 'Quantity',
    type: 'numeric-attribute',
    expression: '[Commerce.Quantity]',
  }),
  Revenue: createAttribute({
    name: 'Revenue',
    type: 'numeric-attribute',
    expression: '[Commerce.Revenue]',
  }),
  VisitID: createAttribute({
    name: 'VisitID',
    type: 'numeric-attribute',
    expression: '[Commerce.Visit ID]',
  }),
  Date: createDateDimension({
    name: 'Date',
    expression: '[Commerce.Date (Calendar)]',
  }),
}) as CommerceDimension;

interface CountryDimension extends Dimension {
  Country: Attribute;
  CountryID: Attribute;
}
export const Country = createDimension({
  name: 'Country',
  Country: createAttribute({
    name: 'Country',
    type: 'text-attribute',
    expression: '[Country.Country]',
  }),
  CountryID: createAttribute({
    name: 'CountryID',
    type: 'numeric-attribute',
    expression: '[Country.Country ID]',
  }),
}) as CountryDimension;
