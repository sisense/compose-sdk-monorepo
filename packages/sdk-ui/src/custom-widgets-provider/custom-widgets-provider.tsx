import { ReactNode, useRef } from 'react';

import { TabberButtonsWidget } from '@/widgets/tabber/tabber-buttons-widget';

import { CustomWidgetsContext } from './custom-widgets-context';
import { CustomWidgetComponent } from './types';

/**
 * Custom Widget Provider component that allows registering and accessing custom widgets.
 *
 * @internal
 */
export const CustomWidgetsProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const customWidgesMapRef = useRef(
    new Map<string, CustomWidgetComponent<any>>([['tabber-buttons', TabberButtonsWidget]]),
  );

  return (
    <CustomWidgetsContext.Provider value={customWidgesMapRef.current}>
      {children}
    </CustomWidgetsContext.Provider>
  );
};
