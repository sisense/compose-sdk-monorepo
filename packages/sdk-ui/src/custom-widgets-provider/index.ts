export * from './custom-widgets-provider';
export * from './custom-widgets-provider-adapter';
export * from './custom-widgets-context';
export * from './types';
export * from './use-execute-custom-widget-query';
export * from './use-custom-widgets';
