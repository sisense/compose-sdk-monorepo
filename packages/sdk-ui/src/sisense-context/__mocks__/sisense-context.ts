export const useSisenseContext = vi.fn();
export const useSisenseContextMock = useSisenseContext;
