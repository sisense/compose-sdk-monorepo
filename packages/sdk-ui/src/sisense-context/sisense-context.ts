import { createContext, useContext } from 'react';
import { ClientApplication } from '../app/client-application';

export type SisenseContextPayload = {
  isInitialized: boolean;
  app?: ClientApplication;
  tracking: {
    enabled: boolean;
    packageName: string;
  };
};

export const SisenseContext = createContext<SisenseContextPayload>({
  isInitialized: false,
  tracking: {
    enabled: true,
    packageName: 'sdk-ui',
  },
});

export const useSisenseContext = () => useContext(SisenseContext);
