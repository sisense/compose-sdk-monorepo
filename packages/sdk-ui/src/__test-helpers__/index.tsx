/**
 * Common test utilities
 */
import { PropsWithChildren, ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import type { Cell, Data } from '@ethings-os/sdk-data';
import isObject from 'lodash-es/isObject';
import userEvent from '@testing-library/user-event';
import { SisenseContext, SisenseContextPayload } from '../sisense-context/sisense-context';
import { Authenticator, HttpClient } from '@ethings-os/sdk-rest-client';
import { ClientApplication } from '../app/client-application';
import { DeepPartial } from 'ts-essentials';
import { useTranslation } from 'react-i18next';
import { translation } from '../translation/resources/en';
import { CustomWidgetsProvider } from '../custom-widgets-provider';
import { MenuProvider } from '@/common/components/menu/menu-provider';

type UserSetupFn = (typeof userEvent)['setup'];
interface SetupResult extends RenderResult {
  user: ReturnType<UserSetupFn>;
}
/**
 * This helper method combines the userEvent.setup() and render() calls at the
 * start of a unit test using @testing-library/react.
 *
 * Based on recommendation by: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
 *
 * @param jsx - React component instance to test, passed directly to render()
 * @param useFakeTimers - true if Vitest fake timers are enabled
 * @returns
 */
export function setup(jsx: ReactElement, useFakeTimers = false): SetupResult {
  return {
    // Need to set "advanceTimers" if we're using Vitest's fake timers.
    // The solution in https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841
    // is for Jest, but this seems to work for Vitest as well.
    user: userEvent.setup(
      useFakeTimers ? { advanceTimers: (ms) => vi.advanceTimersByTime(ms) } : undefined,
    ),
    ...render(jsx),
  };
}

export const MockedHighchartsReact = ({
  options,
}: {
  /**
   * Highcharts chart configuration object.
   * Please refer to the Highcharts (API documentation)[https://api.highcharts.com/highcharts/].
   */
  options: object;
}) => {
  return <div>{JSON.stringify(options)}</div>;
};

type MockedSisenseContextProviderProps = PropsWithChildren<
  DeepPartial<Omit<SisenseContextPayload, 'app'>>
>;

const mockHttpClient = new HttpClient(
  'http://mock-url/sometenant?someparam=true',
  {} as Authenticator,
  'mock-env',
);
export const MockedSisenseContextProvider = ({
  children,
  tracking = {
    enabled: true,
  },
  isInitialized = true,
  errorBoundary,
}: MockedSisenseContextProviderProps) => {
  return (
    <SisenseContext.Provider
      value={{
        app: {
          httpClient: mockHttpClient,
        } as ClientApplication,
        tracking: {
          enabled: tracking.enabled ?? true,
          packageName: tracking.packageName ?? 'sdk-ui',
        },
        isInitialized,
        errorBoundary: { showErrorBox: true, ...errorBoundary },
      }}
    >
      <MenuProvider>
        <CustomWidgetsProvider>{children}</CustomWidgetsProvider>
      </MenuProvider>
    </SisenseContext.Provider>
  );
};

/**
 * Extends the provided data by blurring specified rows.
 *
 * @param {Data} data - The original data object.
 * @param {number[]} blurRowsIndexes - An array of row indexes to be blurred.
 * @returns {Data} The data object with blurred rows.
 */
export const withBlurredRows = (data: Data, blurRowsIndexes: number[]) => ({
  ...data,
  rows: data.rows.map((row, rowIndex) =>
    row.map((value) => {
      const cell: Cell = isObject(value) ? value : { data: value };

      if (blurRowsIndexes.includes(rowIndex)) {
        return { ...cell, blur: true };
      }
      return cell;
    }),
  ),
});

/**
 * Mocking the i18next translation functionality for tests.
 */
export const setupI18nMock = () => {
  vi.mock('react-i18next', async () => {
    const original: { useTranslation: typeof useTranslation } = await vi.importActual(
      'react-i18next',
    );

    // Helper function to get nested translations safely
    const getNestedTranslation = (key: string): string | undefined => {
      return key.split('.').reduce((accumulator: any, currentKey: string) => {
        if (accumulator && typeof accumulator === 'object' && currentKey in accumulator) {
          return accumulator[currentKey];
        }
        return undefined;
      }, translation) as string | undefined;
    };

    return {
      ...original,
      useTranslation: () => ({
        ...original.useTranslation(),
        t: vi.fn((key, options?: Record<string, any>) => {
          const translatedValue = getNestedTranslation(key);
          if (translatedValue && options) {
            // Replace variables in the translated string
            return translatedValue.replace(/{{(.*?)}}/g, (_, varName) => {
              const trimmedVarName = varName.trim();
              return options[trimmedVarName] || `{{${trimmedVarName}}}`; // Fallback if variable is not found
            });
          }
          return translatedValue ?? key; // Fallback to the key if translation is not found
        }),
      }),
    };
  });
};
