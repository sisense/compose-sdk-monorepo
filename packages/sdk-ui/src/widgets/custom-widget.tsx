import { FunctionComponent, useMemo } from 'react';
import { useTranslation } from 'react-i18next';

import { getDataSourceName } from '@sisense/sdk-data';

import { useCustomWidgets } from '@/custom-widgets-provider';
import { CustomWidgetComponentProps } from '@/custom-widgets-provider/types';
import { asSisenseComponent } from '@/decorators/component-decorators/as-sisense-component';
import { withErrorBoundary } from '@/decorators/component-decorators/with-error-boundary';
import { DynamicSizeContainer, getWidgetDefaultSize } from '@/dynamic-size-container';
import ErrorBoundaryBox from '@/error-boundary/error-boundary-box';
import { CustomWidgetProps } from '@/props';
import { useSisenseContext } from '@/sisense-context/sisense-context';

import { WidgetContainer } from './common/widget-container';

const withSize = (props: CustomWidgetComponentProps, size: { width: number; height: number }) => {
  return {
    ...props,
    styleOptions: {
      ...props.styleOptions,
      width: size.width,
      height: size.height,
    },
  };
};

/**
 * Component that renders a custom widget.
 *
 * @internal
 */
export const CustomWidget: FunctionComponent<CustomWidgetProps> = asSisenseComponent({
  componentName: 'CustomWidget',
})((widgetProps: CustomWidgetProps) => {
  const { t } = useTranslation();
  const { getCustomWidget } = useCustomWidgets();
  const { app } = useSisenseContext();

  const renderCustomWidget = getCustomWidget(widgetProps.customWidgetType);
  const renderCustomWidgetWithErrorBoundary = useMemo(() => {
    if (!renderCustomWidget) return null;
    return withErrorBoundary({ componentName: 'CustomWidget' })((props) => (
      <>{renderCustomWidget(props)}</>
    ));
  }, [renderCustomWidget]);

  const dataSource = widgetProps.dataSource ?? app?.defaultDataSource;
  const customWidgetProps: CustomWidgetComponentProps = {
    // TODO - we need to pass dynamically added props, at the same time we wanted to limit amount of the props, so we probably should find a balance here
    ...widgetProps,
    dataSource,
    dataOptions: widgetProps.dataOptions,
    styleOptions: widgetProps.styleOptions,
    filters: widgetProps.filters,
    highlights: widgetProps.highlights,
    description: widgetProps.description,
  };

  const size = useMemo(
    () => ({
      ...(widgetProps.styleOptions?.width ? { width: widgetProps.styleOptions.width } : null),
      ...(widgetProps.styleOptions?.height ? { height: widgetProps.styleOptions.height } : null),
    }),
    [widgetProps.styleOptions?.width, widgetProps.styleOptions?.height],
  );

  const wrapperDefaultSize = useMemo(() => getWidgetDefaultSize('line', { hasHeader: true }), []);
  const chartDefaultSize = useMemo(() => getWidgetDefaultSize('line'), []);

  return (
    <DynamicSizeContainer size={size} defaultSize={wrapperDefaultSize}>
      <WidgetContainer
        title={widgetProps.title}
        description={widgetProps.description}
        styleOptions={widgetProps.styleOptions}
        dataSetName={dataSource && getDataSourceName(dataSource)}
      >
        {renderCustomWidgetWithErrorBoundary ? (
          <DynamicSizeContainer defaultSize={chartDefaultSize}>
            {(size) => renderCustomWidgetWithErrorBoundary(withSize(customWidgetProps, size))}
          </DynamicSizeContainer>
        ) : (
          <ErrorBoundaryBox
            error={t('customWidgets.registerPrompt', {
              customWidgetType: widgetProps.customWidgetType,
            })}
          />
        )}
      </WidgetContainer>
    </DynamicSizeContainer>
  );
});
