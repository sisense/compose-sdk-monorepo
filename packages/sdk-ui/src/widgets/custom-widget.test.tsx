import { render } from '@testing-library/react';

import { CustomWidget } from './custom-widget';
import * as DM from '@/__test-helpers__/sample-ecommerce';

import { useCustomWidgets } from '@/custom-widgets-provider';
import { useSisenseContext } from '@/sisense-context/sisense-context';

vi.mock('@/custom-widgets-provider');
vi.mock('@/sisense-context/sisense-context');

import { Mock } from 'vitest';

describe('CustomWidget', () => {
  const mockUseCustomWidgets = useCustomWidgets as unknown as Mock;
  const mockUseSisenseContext = useSisenseContext as unknown as Mock;

  // We'll store a fake custom widget map in memory
  const customWidgetMap = new Map<string, (props: any) => JSX.Element>();

  beforeEach(() => {
    vi.clearAllMocks();

    // By default, context returns a fallback dataSource
    mockUseSisenseContext.mockReturnValue({
      isInitialized: true,
      app: {
        defaultDataSource: DM.DataSource,
      },
      errorBoundary: {
        showErrorBox: true,
      },
    });

    // By default, no custom widgets in the map; can add in individual tests
    customWidgetMap.clear();
    mockUseCustomWidgets.mockReturnValue({
      registerCustomWidget: vi.fn(),
      hasCustomWidget: vi.fn((key: string) => customWidgetMap.has(key)),
      getCustomWidget: vi.fn((key: string) => customWidgetMap.get(key)),
    });
  });

  it('renders an error if custom widget is not found', () => {
    const { getByRole } = render(
      <CustomWidget customWidgetType="unknown-custom-widget" dataOptions={{}} title={''} />,
    );

    // So let's look for that text in the DOM:
    expect(getByRole('generic', { name: 'error-box' })).toBeInTheDocument();
  });

  it('renders custom widget when custom widget is found', () => {
    const customWidgetComponentMock = vi.fn(() => <div>Custom Widget Content</div>);
    // Add a custom widget to the map that just returns some text
    customWidgetMap.set('my-custom-widget', customWidgetComponentMock);

    const { getByText, queryByRole } = render(
      <CustomWidget
        customWidgetType="my-custom-widget"
        dataOptions={{ someProp: [] }}
        title={'My awesome custom widget'}
        filters={[]}
        highlights={[]}
        styleOptions={{
          someStyle: 'value',
        }}
      />,
    );

    // The error boundary text should not appear
    expect(queryByRole('generic', { name: 'error-box' })).not.toBeInTheDocument();

    // And our custom widget component was called
    expect(customWidgetComponentMock).toHaveBeenCalledOnce();
    expect(customWidgetComponentMock.mock.calls[0]).toMatchSnapshot();

    // Our custom widget text is now rendered
    expect(getByText('Custom Widget Content')).toBeInTheDocument();
  });
});
