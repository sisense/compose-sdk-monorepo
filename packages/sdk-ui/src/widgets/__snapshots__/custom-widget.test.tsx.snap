// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomWidget > renders custom widget when custom widget is found 1`] = `
[
  {
    "customWidgetType": "my-custom-widget",
    "dataOptions": {
      "someProp": [],
    },
    "dataSource": "Sample ECommerce",
    "description": undefined,
    "filters": [],
    "highlights": [],
    "styleOptions": {
      "height": 400,
      "someStyle": "value",
      "width": 400,
    },
    "title": "My awesome custom widget",
  },
]
`;
